lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -27,6 +136,312 @@
+        self.on.routing_table_updated.emit()
+
+    def _router_relation_departed(self, event: RelationDepartedEvent):
+        """If an application has completely departed the relation, remove it
+        from the routing table.
+        """
+        if not self.charm.unit.is_leader():
+            return
+        self._sync_routing_tables()
+
+    def get_routing_table(self) -> RoutingTable:
+        """Build the routing table from all of the related databags. Relations
+        that don't have missing or invalid network requests will be ignored.
+        """
+        logger.debug("Routing Table generation starting.")
+        router_relations = self.model.relations[self.relationship_name]
+        final_routing_table: RoutingTable = {}
+        for relation in router_relations:
+            new_network_name = relation.data[relation.app].get("network-name", None)
+            new_network_request: List[Network] = json.loads(
+                relation.data[relation.app].get("networks", "{}")
+            )
+
+            logger.debug(
+                "Attempting to add (%s) from app:(%s) with relation-name:(%s)",
+                new_network_request,
+                relation.app.name,
+                new_network_name,
+            )
+
+            if (
+                not new_network_name
+                or not new_network_request
+                or _network_name_taken(new_network_name, router_relations)
+            ):
+                continue
+
+            final_routing_table[new_network_name] = []
+
+            for network in new_network_request:
+                try:
+                    _validate_network(network, final_routing_table)
+                except (ValueError, KeyError) as e:
+                    logger.error(
+                        "Exception (%s) occurred with network %s. Skipping this entry.",
+                        e.args[0],
+                        network,
+                    )
+                else:
+                    final_routing_table[new_network_name].append(network)
+
+        logger.debug("Generated rt: %s", final_routing_table)
+        return final_routing_table
+
+    def get_flattened_routing_table(self) -> List[Network]:
+        """Returns a read-only routing table that's flattened to fit the specification.
+
+        Returns:
+            A list of objects of type `Network`
+        """
+        internal_routing_table = self.get_routing_table()
+        final_routing_table: List[Network] = []
+        for networks in internal_routing_table.values():
+            final_routing_table.extend(networks)
+        logger.debug("Flattened RT to %s", final_routing_table)
+        return final_routing_table
+
+    def _sync_routing_tables(self) -> None:
+        """Syncs the internal routing table with all of the requirer's app databags"""
+        routing_table = self.get_flattened_routing_table()
+        logger.info("Resynchronizing routing tables with %s", routing_table)
```

Comment:
Write logs after the action is completed, not before. Here you'd have:

```
logger.info("Resynchronized routing tables with %s", routing_table)
```
after the action was completed

Reply:
Moved most logs to after the action and used past tense